src/ds/sync.rs

```diff
@@ -92,6 +92,155 @@ impl<T> Drop for SpinLockGuard<'_, T> {
     }
 }
 
+pub struct RwSpinLock<T> {
+    lock: AtomicUsize,
+    data: UnsafeCell<T>,
+}
+
+pub struct RwLockWriteGuard<'a, T: 'a> {
+    lock: &'a AtomicUsize,
+    data: &'a mut T,
+}
+
+pub struct RwLockReadGuard<'a, T: 'a> {
+    lock: &'a AtomicUsize,
+    data: &'a T,
+}
+
+const U_MIN: usize = core::isize::MIN as usize;
+
+unsafe impl<T: Send> Send for RwSpinLock<T> {}
+unsafe impl<T: Send + Sync> Sync for RwSpinLock<T> {}
+
+impl<T> RwSpinLock<T> {
+	pub fn new(d: T) -> RwSpinLock<T> {
+		RwSpinLock {
+			lock: AtomicUsize::new(0),
+			data: UnsafeCell::new(d),
+		}
+	}
+	pub fn get_data(self) -> T {
+		let RwSpinLock { data, .. } = self;
+		data.into_inner()
+	}
+
+	pub fn read(&self) -> RwLockReadGuard<T> {
+		while {
+			let mut old_value;
+			while {
+				old_value = self.lock.load(Ordering::Relaxed);
+				old_value & U_MIN != 0
```

Comment:
can you add a quick comment explaining the use of U_MIN in how the lock is implemented